File: app/src/main/java/eu/pkgsoftware/babybuddywidgets/NotesEditorLogic.java
package eu.pkgsoftware.babybuddywidgets;

import android.animation.LayoutTransition;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.ViewGroup;

import eu.pkgsoftware.babybuddywidgets.databinding.NotesEditorBinding;

public class NotesEditorLogic {
    private MainActivity activity;
    private NotesEditorBinding binding;
    private boolean visible;
    private String id = null;
    private CredStore credStore;

    private void updateVisibility() {
        ViewGroup.LayoutParams params = binding.getRoot().getLayoutParams();
        params.height = visible ? ViewGroup.LayoutParams.WRAP_CONTENT : 0;
        params.width = ViewGroup.LayoutParams.MATCH_PARENT;
        binding.getRoot().setLayoutParams(params);
    }

    public NotesEditorLogic(MainActivity activity, NotesEditorBinding binding, boolean visible) {
        this.activity = activity;
        this.visible = visible;
        this.binding = binding;
        credStore = activity.getCredStore();

        updateVisibility();

        LayoutTransition lt = new LayoutTransition();
        lt.setDuration(200);
        lt.enableTransitionType(LayoutTransition.CHANGING);
        lt.enableTransitionType(LayoutTransition.CHANGE_DISAPPEARING);
        lt.enableTransitionType(LayoutTransition.DISAPPEARING);
        //binding.getRoot().setLayoutTransition(lt); - Causes the Notes-dialog to NOT appear. Disabled for now.

        binding.noteEditor.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                credStore.setObjectNotes(id, true, binding.noteEditor.getText().toString());
            }
        });
        binding.noteEditor.setOnFocusChangeListener((v, hasFocus) -> credStore.storePrefs());
    }

    public boolean isVisible() {
        return visible;
    }

    public void setVisible(boolean b) {
        visible = b;
        updateTextFromCredStore();
        updateVisibility();
        if (id != null) {
            credStore.setObjectNotes(id, visible, binding.noteEditor.getText().toString());
            credStore.storePrefs();
        }
    }

    private void updateTextFromCredStore() {
        if (id != null) {
            CredStore.Notes notes = activity.getCredStore().getObjectNotes(id);
            binding.noteEditor.setText(notes.note);
        }
    }

    public void setIdentifier(String id) {
        this.id = id;
        setVisible(credStore.getObjectNotes(id).visible);
        updateTextFromCredStore();
    }

    public String getText() {
        if (visible) {
            return "" + binding.noteEditor.getText();
        } else {
            return "";
        }
    }

    public void clearText() {
        if (id != null) {
            credStore.setObjectNotes(id, visible, null);
            credStore.storePrefs();
        }
        binding.noteEditor.setText("");
    }
}

File: app/src/main/java/eu/pkgsoftware/babybuddywidgets/NotesEditorLogic.kt
package eu.pkgsoftware.babybuddywidgets

import android.animation.LayoutTransition
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.view.View.OnFocusChangeListener
import android.view.ViewGroup
import eu.pkgsoftware.babybuddywidgets.CredStore.Notes
import eu.pkgsoftware.babybuddywidgets.databinding.NotesEditorBinding

interface NotesControl {
    fun getNotes(): Notes
    fun setNotes(notes: Notes)
    fun persistChanges()
}

class NoNotesControl : NotesControl {
    override fun getNotes(): Notes {
        return CredStore.EMPTY_NOTES
    }

    override fun setNotes(notes: Notes) {
    }

    override fun persistChanges() {
    }
}

class CredStoreNotes(val id: String, val credStore: CredStore) : NotesControl {
    override fun getNotes(): Notes {
        return credStore.getObjectNotes(id)
    }

    override fun setNotes(notes: Notes) {
        credStore.setObjectNotes(id, notes.visible, notes.note)
    }

    override fun persistChanges() {
        credStore.storePrefs()
    }
}

class NotesEditorLogic(
    private val binding: NotesEditorBinding,
    private var visible: Boolean
) {
    private var notesControl: NotesControl = NoNotesControl()

    private fun updateVisibility() {
        val params = binding.root.layoutParams
        params.height = if (visible) ViewGroup.LayoutParams.WRAP_CONTENT else 0
        params.width = ViewGroup.LayoutParams.MATCH_PARENT
        binding.root.layoutParams = params
    }

    init {
        updateVisibility()

        val lt = LayoutTransition()
        lt.setDuration(200)
        lt.enableTransitionType(LayoutTransition.CHANGING)
        lt.enableTransitionType(LayoutTransition.CHANGE_DISAPPEARING)
        lt.enableTransitionType(LayoutTransition.DISAPPEARING)
        //binding.getRoot().setLayoutTransition(lt); - Causes the Notes-dialog to NOT appear. Disabled for now.

        binding.noteEditor.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable) {
                notesControl.setNotes(Notes(binding.noteEditor.text.toString(), true))
            }
        })
        binding.noteEditor.onFocusChangeListener =
            OnFocusChangeListener { v: View?, hasFocus: Boolean -> notesControl.persistChanges() }
    }

    fun isVisible(): Boolean {
        return visible
    }

    fun setVisible(b: Boolean) {
        visible = b
        updateTextFromControl()
        updateVisibility()
        notesControl.setNotes(Notes(binding.noteEditor.text.toString(), visible))
        notesControl.persistChanges()
    }

    private fun updateTextFromControl() {
        val notes = notesControl.getNotes()
        binding.noteEditor.setText(notes.note)
    }

    fun setNotes(notesControl: NotesControl) {
        this.notesControl.persistChanges()
        this.notesControl = notesControl
        setVisible(notesControl.getNotes().visible)
        updateTextFromControl()
    }

    val text: String
        get() = if (visible) {
            "" + binding.noteEditor.text
        } else {
            ""
        }

    fun clearText() {
        notesControl.setNotes(CredStore.EMPTY_NOTES)
        notesControl.persistChanges()
        binding.noteEditor.setText("")
    }
}